### ISSUE TITLE
Clarify setup

### ISSUE THREAD

**Issue by seandenigris** [aranega/pybridge#1]
A few questions:
- Why is the python stuff in a separate branch? I don't know much about Python conventions but this seems odd (see https://github.com/aranega/pybridge/tree/python#python-part)
- Is pybridge compatible with others e.g. PythonBridge? I wonder because they both use the `pipenv`
- Is pipenv install done from the image folder or git repo root? I guess the latter is implied from the relative path, but just to confirm...

**Comment by aranega** [comment]
Hi Sean,

Regarding the different branch it's mainly to avoid issues with iceberg. The Python part and the Pharo parts does not evolve exactly at the same time and if I was working all the time on the same branch, dealing with commits from the command line and iceberg always resulted in a _"detached head"_ issue with iceberg, so I decided to shift it to another branch to be able to develop easily. You're right though, it's not really "clean" per say and it would be better to have everything on the same branch, but... technical constraints. I will try to think about a cleaner version though (if you have ideas or suggestions, I'm all for them :D)

Regarding PyBridge and PythonBridge, actually, they are not compatible because they work in a very different way. Funny enough, we developed each other library in parallel without knowing about the other. PythonBridge is more "code generation" oriented while PyBridge is more "object manipulation" oriented and tries to keep a Pharoish way of handling Python object, but it also allows you to deal with Pharo objects in a Pythonish way. Having that (Pharo that can call Python and Python that can call Pharo), then PyBridge is bidirectionnal and is intended to be flexible. The fact PythonBridge and PyBridge are both using `pipenv` is kind of a coincidence. `pipenv` is a way of easily creating virtual environement in Python, which are isolated environement in which you can install all the library you want in a local way (without impacting your full system). It's a common way of working in Python to avoid overriding system packages/libraries or installing an uncompatible version. For each project, you create a virtual env and you install your packages inside of it. For personal uses, I use `pew`for that, but here, I decided to used `pipenv` because it is included in many package managers, unline `pew`. Also, `pipenv` can easily install dependencies from your project (which is the case here while writing `pipenv install`, it install `flask` and `requests`, which are the only dependencies of PyBridge for Python). From my understanding, PythonBridge made the same choice, using `pipenv` to deal with virtual env.

Regarding `pipenv install`, it has to be launched from the git repo root (from the `python` branch), sorry about this, I need to be more clear. 

I will try to improve as much as possible this in the next months, _i.e_: documentation, code, tests... I'm conducting some more experiments at the same time that would allow PyBridge to be more flexible and powerful, but it always requires more time and help :).

**Comment by seandenigris** [comment]
That's all helpful. Thanks.

> "detached head" issue with iceberg

Yes I see how that could be annoying. I guess it's not a big deal, but it makes it a bit harder on the users. E.g. I was getting errors because IIRC I tried to use Iceberg or Gt4Git while the python branch was checked out, and then I had to stop the server to switch back. Hopefully, Iceberg (or GT) will support non-Smalltalk commits a bit better in the near future, but maybe you could create a small script to ease that part of your workflow. Thinking a bit more, I would guess GT would be open to adding it to the git tool and I'm happy to help. The general idea would be to "add all", commit, and then adopt the new commit.

> PyBridge and PythonBridge... are not compatible

Ah, I see, in that you can't e.g. refer to the Python objects of one from the other. Although, they are minimally compatible in the sense that they can both be loaded in the same image without breaking the other. That's good.

> PythonBridge is more "code generation" oriented while PyBridge is more "object manipulation" oriented and tries to keep a Pharoish way of handling Python object, but it also allows you to deal with Pharo objects in a Pythonish way. 

I get it conceptually, but ideally it would be good to have simple examples of what one can/can't or would/wouldn't want to do with PyBridge vs. PythonBridge.

> I decided to used pipenv

pipenv is growing on me as I get used to it. One thing (and I believe PythonBridge makes the same mistake) is that it would be good to run out of a private folder that is *not* part of the git repo - or is ignored by git. The problem is that if I want to add my own dependencies to the Pipfile I make the repo dirty. Maybe a script that copies the files into another location and does `pipenv install` in one step...

**Comment by aranega** [comment]
>  I guess it's not a big deal, but it makes it a bit harder on the users. E.g. I was getting errors because IIRC I tried to use Iceberg or Gt4Git while the python branch was checked out, and then I had to stop the server to switch back. 

I definitely understand. I work with two clones actually, one for Pharo embedded by default inside the image (clone from Pharo using iceberg) and the other clone somewhere else on the `Python` branch. That's the only way I found to be "fluent" with the dev.

> The general idea would be to "add all", commit, and then adopt the new commit.

Something like that would be really helpful. I will check how to do. At the moment, I think iceberg is keeping a kind of "virtual copy" inside the image  that needs to be sync with the repository on the disk (that's why it goes in a "detached head" state so often). With luck the resync that considers external commits will be easy to do.

> I get it conceptually, but ideally it would be good to have simple examples of what one can/can't or would/wouldn't want to do with PyBridge vs. PythonBridge.

Indeed it would be intersting to exhibit differencies. A big part of it will be on the syntax, you use pure Pharo message send with PyBridge and you build the Python AST of the code you want with PythonBridge. I need to dig more also, but I don't think PythonBridge is able to have bidirectionnal communication with Python (I'm not sure at which point you can use Pharo callbacks in the code, or Python callbacks to call Pharo)

>  it would be good to run out of a private folder that is not part of the git repo - or is ignored by git

Definitely, actually, here `pipenv` is only used on my side to be able to create on the fly a quick virtualenv with dependencies for the Python part, but the proper way of doing this would be to release and deploy a pypi package that you install in your own virtualenv. This way, you can choose which virtualenv manager you want to use, you install this package inside and you are good to go. There should be a way also, directly from the image, to be able to reference an existing virtualenv, activate it and launch the server. Another thing that would ease the installation/manipulation would be the automatic creation of a dedicated virtualenv for the image and an easy way of installing dependencies to it from the image if needed. Well, to sum up, it would be interesting to have a Python virtualenv manager inside of the image.
